drivers/ieee802154/ieee802154_dw1000.c

```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
 	return -ENOTSUP;
 }
 
+// From 7.2.31.1 from User Manual
+#define DWT_TX_POWER_CTRL_OFF 7
+#define DWT_TX_POWER_CTRL_COARSE_MASK 0xE0U
+#define DWT_TX_POWER_CTRL_COARSE_STEPS 2.5
+#define DWT_TX_POWER_CTRL_FINE_MASK 0x1FU
+#define DWT_TX_POWER_CTRL_FINE_STEPS 0.5
+
+#define DWT_TX_POWER_TXPOWPHR_SHIFT 8
+#define DWT_TX_POWER_TXPOWSD_SHIFT 16
+
+// These should go into DT probably
+#define DWT_TX_POWER_BOOSTP500_MOD 3
+#define DWT_TX_POWER_BOOSTP250_MOD 6
+#define DWT_TX_POWER_BOOSTP125_MOD 9
+#define DWT_TX_POWER_PHR_MOD 0
+
+uint8_t dwt_encode_power_tx(float power) {
+	uint8_t coarse = DWT_TX_POWER_CTRL_OFF;
+	uint8_t fineTuneVal = 0;
+	if (power >= 0) {
+		coarse--;
+		while (power >= DWT_TX_POWER_CTRL_COARSE_STEPS && coarse > 0) {
+			power -= DWT_TX_POWER_CTRL_COARSE_STEPS;
+			coarse--;
+		}
+		fineTuneVal = (uint8_t) (power / DWT_TX_POWER_CTRL_FINE_STEPS + 0.5) &
+			DWT_TX_POWER_CTRL_FINE_MASK;
+	}
+    uint8_t result = (coarse << 5) | fineTuneVal;
+    return result;
+}
+
 static int dwt_set_power(const struct device *dev, int16_t dbm)
 {
+	if (dbm > 30) {
+		LOG_ERR("Power %u is not available, power must be below 30", dbm);
+		return -EINVAL;
+	}
 	struct dwt_context *ctx = dev->data;
+	uint32_t sys_cfg = dwt_reg_read_u32(dev, DWT_SYS_CFG_ID, 0);
+	uint32_t current_tx_power = dwt_reg_read_u32(dev, DWT_TX_POWER_ID,0);
+	uint32_t tx_power = 0;
 
-	LOG_INF("set_txpower not supported %p", ctx);
-
+	if (sys_cfg & DWT_SYS_CFG_DIS_STXP) {
+		/* SmartPower ON */
+		uint8_t boostp500 = dwt_encode_power_tx(dbm + DWT_TX_POWER_BOOSTP500_MOD);
+		uint8_t boostp250 = dwt_encode_power_tx(dbm + DWT_TX_POWER_BOOSTP250_MOD);
+		uint8_t boostp125 = dwt_encode_power_tx(dbm + DWT_TX_POWER_BOOSTP125_MOD);
+		tx_power = dwt_encode_power_tx(dbm);
+		tx_power |=  ((uint32_t)boostp500 << DWT_TX_POWER_BOOSTP500_SHIFT);
+		tx_power |=  ((uint32_t)boostp250 << DWT_TX_POWER_BOOSTP250_SHIFT);
+		tx_power |=  ((uint32_t)boostp125 << DWT_TX_POWER_BOOSTP125_SHIFT);
+	} else {
+		/* SmartPower OFF */
+		uint8_t tx_pow_phr = dwt_encode_power_tx(dbm + DWT_TX_POWER_PHR_MOD);
```

Comment:
I'd remove this redundant extra calculation in this PR as you're setting PHR_MOD to zero anyway. You can introduce a way to reduce PHR power in a later PR. If you can like you can just add another TODO comment.

Reply:
Outdated